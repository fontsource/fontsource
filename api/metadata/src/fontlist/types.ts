import { type IDResponse } from 'common-api/types';

import type { FontMetadata } from '../types';

const fontlistQueries = [
	'family',
	'subsets',
	'weights',
	'styles',
	'variable',
	'lastModified',
	'category',
	'version',
	'type',
] as const;
type FontlistQueries = (typeof fontlistQueries)[number] & keyof FontMetadata;
const isFontlistQuery = (query: string): query is FontlistQueries =>
	fontlistQueries.includes(query as FontlistQueries);

type Fontlist = Record<string, string | string[] | number[] | boolean>;

type MetadataList = Record<string, IDResponse>;

export { fontlistQueries, isFontlistQuery };
export type { Fontlist, FontlistQueries, MetadataList };
