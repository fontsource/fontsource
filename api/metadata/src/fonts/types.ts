interface ArrayMetadataItem {
	id: string;
	family: string;
	subsets: string[];
	weights: number[];
	styles: string[];
	defSubset: string;
	variable: boolean;
	lastModified: string;
	category: string;
	license: string;
	type: string;
}

type ArrayMetadata = ArrayMetadataItem[];

const fontsQueries = [
	'id',
	'family',
	'subsets',
	'weights',
	'styles',
	'defSubset',
	'variable',
	'lastModified',
	'category',
	'license',
	'type',
] as const;
type FontsQueries = (typeof fontsQueries)[number] & keyof ArrayMetadataItem;
const isFontsQueries = (x: string): x is FontsQueries =>
	fontsQueries.includes(x as FontsQueries);

interface FontVariants {
	[weight: string]: {
		[style: string]: {
			[subset: string]: {
				url: {
					woff2: string;
					woff: string;
					ttf?: string;
					otf?: string;
				};
			};
		};
	};
}

interface IDResponse {
	id: string;
	family: string;
	subsets: string[];
	weights: number[];
	styles: string[];
	defSubset: string;
	variable: boolean;
	lastModified: string;
	category: string;
	license: string;
	type: string;
	unicodeRange: Record<string, string>;
	variants: FontVariants;
}

export type {
	ArrayMetadata,
	ArrayMetadataItem,
	FontsQueries,
	FontVariants,
	IDResponse,
};
export { isFontsQueries };
