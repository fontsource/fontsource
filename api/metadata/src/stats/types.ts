interface StatsResponse {
	npmDownloadTotal: number;
	npmDownloadMonthly: number;
	jsDelivrHitsTotal: number;
	jsDelivrHitsMonthly: number;
}

interface StatsResponseAll {
	total: StatsResponse;
	static: StatsResponse;
	variable?: StatsResponse;
}

type StatsResponseAllRecord = Record<string, StatsResponseAll>;

export type { StatsResponse, StatsResponseAll, StatsResponseAllRecord };
