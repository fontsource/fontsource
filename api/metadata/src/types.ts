type CFRouterContext = [env: Env, ctx: ExecutionContext];

interface VariableMetadata {
	[axes: string]: {
		default: string;
		min: string;
		max: string;
		step: string;
	};
}

interface LicenseMetadata {
	type: string;
	url: string;
	attribution: string;
}

interface FontMetadata {
	id: string;
	family: string;
	subsets: string[];
	weights: number[];
	styles: string[];
	defSubset: string;
	variable: false | VariableMetadata;
	lastModified: string;
	category: string;
	license: LicenseMetadata;
	source: string;
	type: 'google' | 'icons' | 'other';
	version: string;
	npmVersion: string;
	unicodeRange: {
		[subset: string]: string;
	};
}

type FontsourceMetadata = Record<string, FontMetadata>;

export type { CFRouterContext, FontsourceMetadata, FontMetadata };
