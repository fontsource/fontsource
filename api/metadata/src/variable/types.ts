interface AxesData {
	default: string;
	min: string;
	max: string;
	step: string;
}

// axes: italic: link
type VariableVariants = Record<string, Record<string, string>>;

interface VariableMetadata {
	family: string;
	id: string;
	axes: Record<string, AxesData>;
}

interface VariableMetadataWithVariants extends VariableMetadata {
	variants: VariableVariants;
}

type VariableList = Record<string, VariableMetadata>;

interface AxisRegistryItem {
	name: string;
	tag: string;
	description: string;
	min: number;
	max: number;
	default: number;
	precision: number;
}

type AxisRegistryDownload = AxisRegistryItem[];
type AxisRegistry = Record<string, Omit<AxisRegistryItem, 'tag'>>;

const axisRegistryQueries = ['name', 'tag'] as const;
type AxisRegistryQueries = (typeof axisRegistryQueries)[number];
const isAxisRegistryQuery = (x: string): x is AxisRegistryQueries =>
	axisRegistryQueries.includes(x as AxisRegistryQueries);

export type {
	AxisRegistry,
	AxisRegistryDownload,
	VariableList,
	VariableMetadata,
	VariableMetadataWithVariants,
};

export { isAxisRegistryQuery };
