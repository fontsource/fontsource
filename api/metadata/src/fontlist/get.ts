import type { FontsourceMetadata, TTLMetadata } from '../types';
import type { Fontlist, FontlistQueries } from './types';
import { updateList, updateMetadata } from './update';

const getOrUpdateMetadata = async (
	env: Env,
	ctx: ExecutionContext,
): Promise<FontsourceMetadata> => {
	const { value, metadata } = await env.FONTLIST.getWithMetadata<
		FontsourceMetadata,
		TTLMetadata
	>('metadata', {
		type: 'json',
	});

	if (!value) {
		return await updateMetadata(env);
	}

	// If the ttl is not set or the cache expiry is less than the current time, then return old value
	// while revalidating the cache
	if (!metadata?.ttl || metadata.ttl < Date.now() / 1000) {
		ctx.waitUntil(updateMetadata(env));
	}

	return value;
};

const getOrUpdateList = async (
	key: FontlistQueries,
	env: Env,
	ctx: ExecutionContext,
): Promise<Fontlist> => {
	const { value, metadata } = await env.FONTLIST.getWithMetadata<
		Fontlist,
		TTLMetadata
	>(key, {
		type: 'json',
	});

	if (!value) {
		return await updateList(key, env);
	}

	if (!metadata?.ttl || metadata.ttl < Date.now() / 1000) {
		ctx.waitUntil(updateList(key, env));
	}

	return value;
};

export { getOrUpdateList, getOrUpdateMetadata };
