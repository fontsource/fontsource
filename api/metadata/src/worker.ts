import fontsRouter from './fonts/router';
import fontlistRouter from './fontlist/router';
import downloadRouter from './download/router';
import { error } from 'itty-router';

export default {
	async fetch(
		request: Request,
		env: Env,
		ctx: ExecutionContext
	): Promise<Response> {
		try {
			const url = new URL(request.url);

			if (url.pathname.startsWith('/fontlist')) {
				return fontlistRouter.handle(request, env, ctx).catch((err) => {
					console.error(err);
					return error(err);
				});
			}

			if (url.pathname.startsWith('/v1/fonts')) {
				return fontsRouter.handle(request, env, ctx).catch((err) => {
					console.error(err);
					return error(err);
				});
			}

			if (url.pathname.startsWith('/v1/download')) {
				return downloadRouter.handle(request, env, ctx).catch((err) => {
					console.error(err);
					return error(err);
				});
			}

			return error(
				404,
				'Not Found. Please refer to the Fontsource API documentation: https://fontsource.org/docs/api'
			);
		} catch (e) {
			console.error(e);
			return error(500, 'Internal Server Error.');
		}
	},
};
