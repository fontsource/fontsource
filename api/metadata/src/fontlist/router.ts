import { error, type IRequestStrict, json, Router } from 'itty-router';

import type { CFRouterContext } from '../types';
import { CF_EDGE_TTL } from '../utils';
import { getOrUpdateList } from './get';
import { type Fontlist, isFontlistQuery } from './types';

const router = Router<IRequestStrict, CFRouterContext>();

router.get('/fontlist', async (request, env, ctx) => {
	// Get query string
	const url = new URL(request.url);
	const queryString = url.searchParams.toString();

	// If there is more than 2 query strings, then return 400
	if (queryString.split('&').length >= 2) {
		return error(400, 'Bad Request. You can only use one query parameter.');
	}

	// If there is no query string, then return the type list
	let list: Fontlist | undefined;
	if (queryString.length === 0) {
		list = await getOrUpdateList('type', env, ctx);
	}

	// If there is a query string, then return the respective list
	if (queryString.length > 0) {
		// Get query string key
		const query = url.searchParams.keys().next().value;

		// Type guard
		if (!isFontlistQuery(query)) {
			return error(400, 'Bad Request. Invalid query parameter.');
		}

		// Get or update the list
		list = await getOrUpdateList(query, env, ctx);
	}

	if (!list) {
		return error(500, 'Internal server error.');
	}

	return json(list, {
		headers: {
			'CDN-Cache-Control': `max-age=${CF_EDGE_TTL}`,
		},
	});
});

// 404 for everything else
router.all('*', () =>
	error(
		404,
		'Not Found. Please refer to the Fontsource API documentation: https://fontsource.org/docs/api',
	),
);

export default router;
