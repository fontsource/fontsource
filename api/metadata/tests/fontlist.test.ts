import { beforeEach, expect, it } from 'vitest';

import { fontlistQueries } from '../src/fontlist/types';
import worker from '../src/worker';
import { mockMetadata } from './helpers';

const describe = setupMiniflareIsolatedStorage();

describe('fontlist worker', () => {
	const env = getMiniflareBindings() satisfies Env;
	const ctx = new ExecutionContext();

	beforeEach(async () => {
		mockMetadata();
	});

	it('should return type as default request', async () => {
		const request = new Request('http://localhost:8787/fontlist');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(200);

		const body = await response.json();
		expect(body).toEqual({
			akshar: 'google',
			alegreya: 'google',
			archivo: 'google',
			ballet: 'google',
			fraunces: 'google',
			'material-icons': 'google',
			'material-symbols-outlined': 'google',
			'noto-sans-jp': 'google',
			'noto-serif-hk': 'google',
			recursive: 'google',
			'roboto-flex': 'google',
			yakuhanjp: 'other',
		});
	});

	const queryParams = fontlistQueries;
	for (const query of queryParams) {
		it(`should return return value of ${query} query string`, async () => {
			const request = new Request(`http://localhost:8787/fontlist?${query}`);
			const response = await worker.fetch(request, env, ctx);
			expect(response.status).toBe(200);

			const body = await response.json();
			expect(body).toMatchSnapshot();
		});
	}

	it('should return 400 for invalid query string', async () => {
		const request = new Request('http://localhost:8787/fontlist?invalid');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(400);
	});

	it('should return 400 for multiple query strings', async () => {
		const request = new Request('http://localhost:8787/fontlist?type&category');
		const response = await worker.fetch(request, env, ctx);
		expect(response.status).toBe(400);
	});
});
