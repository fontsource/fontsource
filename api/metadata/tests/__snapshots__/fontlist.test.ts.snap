// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fontlist worker > should return return value of category query string 1`] = `
{
  "akshar": "sans-serif",
  "alegreya": "serif",
  "archivo": "sans-serif",
  "ballet": "handwriting",
  "fraunces": "serif",
  "material-icons": "icons",
  "material-symbols-outlined": "icons",
  "noto-sans-jp": "sans-serif",
  "noto-serif-hk": "serif",
  "recursive": "sans-serif",
  "roboto-flex": "sans-serif",
  "yakuhanjp": "other",
}
`;

exports[`fontlist worker > should return return value of family query string 1`] = `
{
  "akshar": "Akshar",
  "alegreya": "Alegreya",
  "archivo": "Archivo",
  "ballet": "Ballet",
  "fraunces": "Fraunces",
  "material-icons": "Material Icons",
  "material-symbols-outlined": "Material Symbols Outlined",
  "noto-sans-jp": "Noto Sans JP",
  "noto-serif-hk": "Noto Serif HK",
  "recursive": "Recursive",
  "roboto-flex": "Roboto Flex",
  "yakuhanjp": "YakuHanJP",
}
`;

exports[`fontlist worker > should return return value of lastModified query string 1`] = `
{
  "akshar": "2023-03-21",
  "alegreya": "2023-03-21",
  "archivo": "2022-09-22",
  "ballet": "2023-03-09",
  "fraunces": "2023-03-21",
  "material-icons": "2023-03-08",
  "material-symbols-outlined": "2023-04-27",
  "noto-sans-jp": "2023-04-13",
  "noto-serif-hk": "2022-09-27",
  "recursive": "2023-01-06",
  "roboto-flex": "2023-01-06",
  "yakuhanjp": "2023-05-15",
}
`;

exports[`fontlist worker > should return return value of styles query string 1`] = `
{
  "akshar": [
    "normal",
  ],
  "alegreya": [
    "italic",
    "normal",
  ],
  "archivo": [
    "italic",
    "normal",
  ],
  "ballet": [
    "normal",
  ],
  "fraunces": [
    "italic",
    "normal",
  ],
  "material-icons": [
    "normal",
  ],
  "material-symbols-outlined": [
    "normal",
  ],
  "noto-sans-jp": [
    "normal",
  ],
  "noto-serif-hk": [
    "normal",
  ],
  "recursive": [
    "normal",
  ],
  "roboto-flex": [
    "normal",
  ],
  "yakuhanjp": [
    "normal",
  ],
}
`;

exports[`fontlist worker > should return return value of subsets query string 1`] = `
{
  "akshar": [
    "devanagari",
    "latin",
    "latin-ext",
  ],
  "alegreya": [
    "cyrillic",
    "cyrillic-ext",
    "greek",
    "greek-ext",
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "archivo": [
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "ballet": [
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "fraunces": [
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "material-icons": [
    "latin",
  ],
  "material-symbols-outlined": [
    "latin",
  ],
  "noto-sans-jp": [
    "cyrillic",
    "japanese",
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "noto-serif-hk": [
    "chinese-hongkong",
    "cyrillic",
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "recursive": [
    "cyrillic-ext",
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "roboto-flex": [
    "cyrillic",
    "cyrillic-ext",
    "greek",
    "latin",
    "latin-ext",
    "vietnamese",
  ],
  "yakuhanjp": [
    "latin",
  ],
}
`;

exports[`fontlist worker > should return return value of type query string 1`] = `
{
  "akshar": "google",
  "alegreya": "google",
  "archivo": "google",
  "ballet": "google",
  "fraunces": "google",
  "material-icons": "google",
  "material-symbols-outlined": "google",
  "noto-sans-jp": "google",
  "noto-serif-hk": "google",
  "recursive": "google",
  "roboto-flex": "google",
  "yakuhanjp": "other",
}
`;

exports[`fontlist worker > should return return value of variable query string 1`] = `
{
  "akshar": true,
  "alegreya": true,
  "archivo": true,
  "ballet": true,
  "fraunces": true,
  "material-icons": false,
  "material-symbols-outlined": true,
  "noto-sans-jp": true,
  "noto-serif-hk": true,
  "recursive": true,
  "roboto-flex": true,
  "yakuhanjp": false,
}
`;

exports[`fontlist worker > should return return value of version query string 1`] = `
{
  "akshar": "v9",
  "alegreya": "v35",
  "archivo": "v18",
  "ballet": "v25",
  "fraunces": "v30",
  "material-icons": "v140",
  "material-symbols-outlined": "v108",
  "noto-sans-jp": "v51",
  "noto-serif-hk": "v2",
  "recursive": "v37",
  "roboto-flex": "v9",
  "yakuhanjp": "v3.3.1",
}
`;

exports[`fontlist worker > should return return value of weights query string 1`] = `
{
  "akshar": [
    300,
    400,
    500,
    600,
    700,
  ],
  "alegreya": [
    400,
    500,
    600,
    700,
    800,
    900,
  ],
  "archivo": [
    100,
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
  ],
  "ballet": [
    400,
  ],
  "fraunces": [
    100,
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
  ],
  "material-icons": [
    400,
  ],
  "material-symbols-outlined": [
    100,
    200,
    300,
    400,
    500,
    600,
    700,
  ],
  "noto-sans-jp": [
    100,
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
  ],
  "noto-serif-hk": [
    200,
    300,
    400,
    500,
    600,
    700,
    800,
    900,
  ],
  "recursive": [
    300,
    400,
    500,
    600,
    700,
    800,
    900,
  ],
  "roboto-flex": [
    400,
  ],
  "yakuhanjp": [
    100,
    200,
    300,
    400,
    500,
    700,
    900,
  ],
}
`;
